### PR TITLE
Prototype code for seeing if FindPython3 is usable for rosidl_python

### PR DESCRIPTION
**Pull request by sloretz**
This is a prototype intended to see how usable FindPython3 on non-linux platforms is. `rosidl_generator_py` makes the most use of `FindPythonExtra`.

Should inform https://github.com/ros2/python_cmake_module/pull/7

### CODE REVIEW

**Comment by sloretz** [comment]
Full CI with build type None

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14973/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9714/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12653/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15137/)

Full CI with build type Release

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14974/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9715/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12654/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15138/)

Not doing Debug, because I would expect this to fail on windows as it's currently written

**Comment by sloretz** [comment]
Full CI with build type None

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/14983/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/9724/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/12669/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/15147/)

**Comment by ivanpauno** [comment]
@sloretz I'm curious what were the issues with this, do you remember?
The CI jobs are too old so I cannot see the logs. Thanks!

**Comment by clalancette** [comment]
CI for this is over in https://github.com/ros2/python_cmake_module/pull/7#issuecomment-1877644112

File: rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake [hunk 1]
```diff
@@ -17,10 +17,19 @@ find_package(rosidl_runtime_c REQUIRED)
 find_package(rosidl_typesupport_c REQUIRED)
 find_package(rosidl_typesupport_interface REQUIRED)
 
-find_package(PythonInterp 3.6 REQUIRED)
-
-find_package(python_cmake_module REQUIRED)
-find_package(PythonExtra MODULE REQUIRED)
+find_package(Python3 REQUIRED COMPONENTS Interpreter Development)
+if(WIN32 AND CMAKE_BUILD_TYPE STREQUAL "Debug")
```
Comment (sloretz) [inline rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake:21]: IIUC it looks like this is using the path to the debug executable as a hint to find the debug version of Numpy, but in doing so it duplicates the logic in PythonExtra to find the debug version of Python.

One option, I see https://github.com/ros2/python_cmake_module/pull/7 references a `PythonExtra_EXECUTABLE` variable. How about we promote that to an output variable so that code here could set the `Python3_EXECUTABLE` hint to it's value?

Another option, what if `PythonExtra` supported components, such that it passed the given components on to its own `find_package(Python3 ...)` call?

File: rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake [hunk 2]
```diff
@@ -142,12 +146,9 @@ set_property(
   ${_generated_extension_files} ${_generated_py_files} ${_generated_c_files}
   PROPERTY GENERATED 1)
 
-# Export target so downstream interface packages can link to it
-set(rosidl_generator_py_suffix "__rosidl_generator_py")
-
-set(_target_name_lib "${rosidl_generate_interfaces_TARGET}${rosidl_generator_py_suffix}")
+set(_target_name_lib "${rosidl_generate_interfaces_TARGET}__python")
```
Comment (sloretz) [inline rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake:149]: I notice this changes the target suffix from `__rosidl_generator_py` to `__python`, which is fine. Just curious was this change necessary or just making it look nicer?

File: rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake (hunk 1 above)
Comment (clalancette) [reply rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake]: I did something slightly different.  I ended up reinstating the `find_package` calls to both `python_cmake_module` and `PythonExtra`, and then fetching the target property `Python3_EXECUTABLE` from `PythonExtra::Interpreter`.  I believe that should have the desired effect, while still reusing the logic from `python_cmake_module`.

File: rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake [hunk 3]
```diff
@@ -156,44 +148,23 @@ add_dependencies(
...
-  message(STATUS "Using numpy include directory: ${_output}")
-  target_include_directories(${_target_name_lib} PUBLIC "${_output}")
+set(_extension_compile_flags "")
+if(CMAKE_COMPILER_IS_GNUCXX OR CMAKE_CXX_COMPILER_ID MATCHES "Clang")
+  set(_extension_compile_flags -Wall -Wextra)
 endif()
 
 rosidl_get_typesupport_target(c_typesupport_target "${rosidl_generate_interfaces_TARGET}" "rosidl_typesupport_c")
-target_link_libraries(${_target_name_lib} ${c_typesupport_target})
+target_link_libraries(${_target_name_lib} PRIVATE ${c_typesupport_target})
```
Comment (sloretz) [inline rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake:167]: I can't quite figure it out in my mind, but setting the c type support library dependency as private seems like it could have some relationship to the Windows Debug test failures:

https://ci.ros2.org/job/ci_windows/20772/testReport/junit/(root)/projectroot/test_publisher_subscriber__rclpy__rmw_connextdds/

```
[test_subscriber-3] Traceback (most recent call last):
[test_subscriber-3]   File "C:\ci\ws\install\Lib\site-packages\rosidl_generator_py\import_type_support_impl.py", line 46, in import_type_support
[test_subscriber-3]     return importlib.import_module(module_name, package=pkg_name)
[test_subscriber-3]   File "C:\Python38\lib\importlib\__init__.py", line 127, in import_module
[test_subscriber-3]     return _bootstrap._gcd_import(name[level:], package, level)
[test_subscriber-3]   File "<frozen importlib._bootstrap>", line 1014, in _gcd_import
[test_subscriber-3]   File "<frozen importlib._bootstrap>", line 991, in _find_and_load
[test_subscriber-3]   File "<frozen importlib._bootstrap>", line 973, in _find_and_load_unlocked
[test_publisher-2]     check_is_valid_msg_type(msg_type)
[test_publisher-2]   File "C:\ci\ws\install\Lib\site-packages\rclpy\type_support.py", line 35, in check_is_valid_msg_type
[test_publisher-2]     check_for_type_support(msg_type)
[test_publisher-2]   File "C:\ci\ws\install\Lib\site-packages\rclpy\type_support.py", line 29, in check_for_type_support
[test_publisher-2]     msg_or_srv_type.__class__.__import_type_support__()
[test_publisher-2]   File "C:\ci\ws\install\Lib\site-packages\rcl_interfaces\msg\_parameter_event.py", line 36, in __import_type_support__
[test_publisher-2]     module = import_type_support('rcl_interfaces')
[test_publisher-2]   File "C:\ci\ws\install\Lib\site-packages\rosidl_generator_py\import_type_support_impl.py", line
```

Comment (sloretz) [reply rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake:167]: Or maybe the first place to look is to ask what is `import_type_support()` doing?

Comment (clalancette) [reply rosidl_generator_py/cmake/rosidl_generator_py_generate_interfaces.cmake:167]: I think we solved this with `set_target_properties(${_target_name} PROPERTIES DEBUG_POSTFIX "${PythonExtra_POSTFIX}")`.  I'm going to resolve this for now, and I'll keep an eye on the CI at https://github.com/ros2/python_cmake_module/pull/7#issuecomment-1912646583

**Review by sloretz — COMMENTED** [review]
LGTM!

**Review by clalancette — APPROVED** [review]

**Comment by clalancette** [comment]
CI for this is over in https://github.com/ros2/python_cmake_module/pull/7#issuecomment-1937404999
